src/lib.rs

```diff
@@ -371,38 +346,23 @@ pub(crate) unsafe fn rav1d_open(c_out: &mut *mut Rav1dContext, s: &Rav1dSettings
         rav1d_refmvs_init(&mut (*f).rf);
         n = n.wrapping_add(1);
     }
-    let mut m: c_uint = 0 as c_int as c_uint;
-    while m < (*c).n_tc {
-        let t: *mut Rav1dTaskContext = &mut *((*c).tc).offset(m as isize) as *mut Rav1dTaskContext;
-        (*t).f = &mut *((*c).fc).offset(0) as *mut Rav1dFrameContext;
+    for _ in 0..(*c).n_tc {
         let task_thread = Arc::new(Rav1dTaskContext_task_thread::new(Arc::clone(
             &(*c).task_thread,
         )));
-        (&mut (*t).task_thread as *mut Arc<Rav1dTaskContext_task_thread>)
-            .write(Arc::clone(&task_thread));
-        *BitDepth16::select_mut(&mut (*t).cf) = Align64([0; 32 * 32]);
-        let thread = if (*c).n_tc > 1 as c_uint {
-            let thread_args = Box::new(Rav1dTaskContext_borrow {
-                c: &*c,
-                tc: &mut *t,
-            });
-            let mut thread = 0;
-            if pthread_create(
-                &mut thread,
-                &mut thread_attr,
-                Some(rav1d_worker_task),
-                Box::into_raw(thread_args).cast(),
-            ) != 0
-            {
-                return error(c, c_out, &mut thread_attr);
-            }
-            Some(thread)
+        if (*c).n_tc > 1 {
+            let context_borrow = &*c;
+            let task_thread_copy = Arc::clone(&task_thread);
+            let handle = thread::spawn(|| rav1d_worker_task(context_borrow, task_thread_copy));
+            (*c).tc
+                .push(Rav1dContextTaskThread::new(Some(handle), task_thread));
         } else {
-            None
-        };
-        (*c).tc_shared
-            .push(Rav1dContextTaskThread::new(thread, task_thread));
-        m = m.wrapping_add(1);
+            (*c).main_tc = Some(Mutex::new(Box::new(Rav1dTaskContext::new(
+                &*((*c).fc).offset(0),
+                Arc::clone(&task_thread),
+            ))));
+            (*c).tc.push(Rav1dContextTaskThread::new(None, task_thread));
```

Comment:
The C allocated all up front, so this should either be a `(0..(*c).n_fc).map(|_| ...).collect()` or there should be a `Vec::with_capacity` at the beginning.

Reply:
I don't think we care, this is in a small loop and will get initialized together. This structure didn't even exist in the original, it's new. Trying to optimize its allocation seems like premature optimization.

Reply:
I dropped it into a boxed slice just to make it clear the size doesn't change after initialization.

Reply:
Oh perfect, I was just gonna suggest that, too.